Woodoku/src/Entity/WoodokuBoard.py

```diff
@@ -25,89 +25,191 @@ def remove_blocks(self, blocks_coord: List[Tuple[int, int]]) -> None:
+        return true. Otherwise, false.
+
+        Args:
+            blocks_coord (List[Tuple[int, int]]): list of blocks to check
+
+        Returns:
+            bool: if all blocks in `blocks_coord` is empty
+        """
+    
+    def __validate(block: Tuple[int, int]) -> None:
+        """validate if block is within the 9x9 board. raise Error if not.
+
+        Args:
+            block (Tuple[int, int]): The block to be validated
+
+        Raises:
+            ShapeOutOfBoardError: when some block is not valid 
         """
         pass
+    
+    
+    def __str__(self) -> str:
+        pass
 
 
 class WoodokuBoard:
+    """A 9x9 Woodoku Board"""
+
     __scoreAgent: ScoreAgent
-    __representation: WoodokuBoardRepresentation
+    __representation: _WoodokuBoardRepresentation
 
     def __init__(self):
-        pass
-
-    def __str__(self):
-        pass
+        self.__scoreAgent = ScoreAgent()
+        self.__representation = _WoodokuBoardRepresentation()
 
-    def initialize(self):
-        pass
 
     def can_add_shape_to_board(self, shape: WoodokuShape) -> bool:
+        """Check if the woodoku shape can fit into the board. If all current
+        shapes are not be able to add, the game fails.
+
+        scans the board in a random fashion to average the complexity
+
+        Args:
+            shape (WoodokuShape): a woodoku shape for validation
+
+        Returns:
+            bool: if the `shape` fits the board
         """
-        Check if the woodoku shape can fit into the board. If all current shapes are not be able to add, the game fails.
-        :return: if can fit.
-        """
-        pass
+        rng = np.random.default_rng()
+        all_blocks = [(i, j) for i in range(9) for j in range(9)]
+        rng.shuffle(all_blocks)
+
+        for row, col in all_blocks:
+            if self.can_add_shape_at_location(shape, x=row, y=col):
+                return True
+
+        return False
 
     def can_add_shape_at_location(self, shape: WoodokuShape, x: int, y: int) -> bool:
+        """Check if the woodoku `shape` can be placed into the board at location
+        `(x, y)` coordinates as the left upper corner of the shape.
+
+        Args:
+            shape (WoodokuShape): The shape needed to be checked
+            x (int): x coordinate
+            y (int): y coordinate
+
+        Returns:
+            bool: if `shape` can be added to `(x,y)`
         """
-        Check if the woodoku shape can be placed into the board at location (x, y) coordinates as the left upper corner
-        of the shape.
-        :param shape: the shape
-        :param x: x coordinate
-        :param y: y coordinate
-        :return: if can be placed at location.
-        """
-        pass
+        blocks = shape.map_to_board_at(x,y)
+        try:
+            return self.__representation.is_not_occupied(blocks)
+        except ShapeOutOfBoardError:
+            return False
```

Comment:
The upper level game logic might still need a difference between that a shape cannot be added at location and the coordinates were incorrect. Is there any particular reason why we are not throwing this error?

Reply:
I agree, the only scenario for handling this error that I can think of is for printing the error to `CommandLineUI` . For now, I'll let it throw